### ISSUE TITLE
Commander unsafely allows commands with duplicate `name`s to be registered

### ISSUE THREAD

**Issue by huntie** [tj/commander.js#1903]
We ran into an issue recently in a project that depends on `commander`, which dynamically defines commands based on plugins. Commander's internal array data structure for commands, and the `Array.find` based lookup on `name`, meant that repeat calls to `command()` produced unexpected behaviour (the most recent call does **_not_** redeclare the command).

At minimum, I believe this should **error**.

- https://github.com/react-native-community/cli/pull/1999

**Example**

```js
const { Command } = require('commander');
const program = new Command();
program.command('split').action((str, options) => console.log('1st'));
program.command('split').action((str, options) => console.log('2nd'));
program.parse();
```
Output will always be the first.
```sh
$ node example
Usage: example [options] [command]
Options:
  -h, --help      display help for command
Commands:
  split
  split
  help [command]  display help for command
$ node example split
1st
```

**Expectation**

Shouldn't allow duplicates and should throw an exception if a `name` already exists.

**Comment by shadowspawn** [comment]
Commander does not have tests for all possible mistakes/accidental-misuse by the "author", but willing to consider more as they come up. Mixing in plugins does make it possible for the issue to come up for the end-user rather than for the author as such.

The duplicate primary name does have a dubious outcome. I am less worried about aliases clashing, but could perhaps actually handle this and drop clashing alias.

**Comment by aweebit** [comment]
Opened #1924 with a fix.

**Comment by shadowspawn** [comment]
Reworked PR in #2509.

**Comment by shadowspawn** [comment]
Released in Commander v12.0.0.
